app/src/main/java/com/arctouch/codechallenge/Controller.java

```diff
@@ -0,0 +1,126 @@
+package com.arctouch.codechallenge;
+
+import android.annotation.SuppressLint;
+import android.app.Application;
+import android.text.TextUtils;
+import android.widget.Toast;
+
+import com.arctouch.codechallenge.api.TmdbApi;
+import com.arctouch.codechallenge.base.MovieListReceiver;
+import com.arctouch.codechallenge.dao.GenreDao;
+import com.arctouch.codechallenge.dao.MovieDao;
+import com.arctouch.codechallenge.data.Cache;
+import com.arctouch.codechallenge.model.Genre;
+import com.arctouch.codechallenge.model.Movie;
+import com.arctouch.codechallenge.model.UpcomingMoviesResponse;
+import com.arctouch.codechallenge.util.NetworkObserver;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import io.reactivex.android.schedulers.AndroidSchedulers;
+import io.reactivex.schedulers.Schedulers;
+import io.realm.RealmList;
+import okhttp3.OkHttpClient;
+import retrofit2.Retrofit;
+import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
+import retrofit2.converter.moshi.MoshiConverterFactory;
+
+public class Controller extends Application{
+
+    private static Controller instance;
```

Comment:
Are the problems involved in having `static` references to variables of type `Context`?

Reply:
This `Controller` class was intended to be a Singleton, that's why I used a static reference to its context, but the problem with it is to have memory leaks. I used it because I thought that a static reference to an application context wouldn’t cause a memory leak since the application context is around for the lifetime the app is running anyway. But, in the end, it was an error as it's a encapsulation problem.